libsel4vmmplatsupport/src/drivers/virtio_emul.c

```diff
@@ -86,7 +86,7 @@ static int emul_io_out(virtio_emul_t *emul, unsigned int offset, unsigned int si
         assert(size == 4);
         int queue = emul->virtq.queue;
         emul->virtq.queue_pfn[queue] = value;
-        vring_init(&emul->virtq.vring[queue], emul->virtq.queue_size[queue], (void *)(uintptr_t)(value << 12),
+        vring_init(&emul->virtq.vring[queue], emul->virtq.queue_size[queue], (void *)((uintptr_t)value << 12),
```

Comment:
Maybe adding a check (or at least an assert) would be good to ensure the shift by 12 does not overflow. For 64-bit this seem quite unlikely, but for 32-bit it might still happen.

Reply:
if (4 = sizeof(uintptr_t)) {
    assert(value < BIT(20));
}

Your mean like this?

Reply:
Well，although it lookes like generic, the value notified by Host with iowrite32 is unsigned int type, the valid range is less than BIT(32) even for 64-bit, so value >= BIT((8 * sizeof(uintptr_t)) - 12) is never ture

Reply:
I don't think the assert is necessary, given the sizes and types of the operands.  It may be worth scanning the code for other cases where the cast is in the wrong place.